src/node_lttng.cc

```diff
@@ -0,0 +1,264 @@
+#include "util.h"
+
+#ifdef HAVE_LTTNG
+#include "node_lttng.h"
+#include "node_lttng_provider.h"
+#include <string.h>
+#else
+#define NODE_HTTP_SERVER_REQUEST(arg0, arg1)
+#define NODE_HTTP_SERVER_REQUEST_ENABLED() (0)
+#define NODE_HTTP_SERVER_RESPONSE(arg0)
+#define NODE_HTTP_SERVER_RESPONSE_ENABLED() (0)
+#define NODE_HTTP_CLIENT_REQUEST(arg0, arg1)
+#define NODE_HTTP_CLIENT_REQUEST_ENABLED() (0)
+#define NODE_HTTP_CLIENT_RESPONSE(arg0)
+#define NODE_HTTP_CLIENT_RESPONSE_ENABLED() (0)
+#define NODE_NET_SERVER_CONNECTION(arg0)
+#define NODE_NET_SERVER_CONNECTION_ENABLED() (0)
+#define NODE_NET_STREAM_END(arg0)
+#define NODE_NET_STREAM_END_ENABLED() (0)
+#define NODE_GC_START(arg0, arg1, arg2)
+#define NODE_GC_DONE(arg0, arg1, arg2)
+#endif
+
+#include "env.h"
+#include "env-inl.h"
+
+namespace node {
+
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::GCCallbackFlags;
+using v8::GCEpilogueCallback;
+using v8::GCPrologueCallback;
+using v8::GCType;
+using v8::Handle;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Value;
+
+#define SLURP_STRING(obj, member, valp) \
+  if (!(obj)->IsObject()) { \
+    return env->ThrowError( \
+        "expected object for " #obj " to contain string member " #member); \
+  } \
+  node::Utf8Value _##member(env->isolate(), \
+      obj->Get(OneByteString(env->isolate(), #member))); \
+  if ((*(const char **)valp = *_##member) == nullptr) \
+    *(const char **)valp = "<unknown>";
```

Comment:
I speculate that the dtrace code was written by someone who is more familiar with C than C++.  The things that are wrong with the correct approach are:

1. It violates const correctness.  In theory, that can lead to the compiler emitting bad code because certain preconditions don't hold.

2. Assigning a read-only string to a mutable pointer.  If the program actually tries to write to the read-only string, it will in likelihood die with a SIGSEGV or a SIGBUS.  It doesn't do that right now but it's a bug waiting to happen.

If you need to mutate the string, do it like this:

```c++
Local<Value> val = obj->Get(...);
if (val->IsString()) {
  Local<String> str(val.As<String>());
  char buffer[512];  // or whatever size is appropriate
  str->WriteUtf8(buffer, sizeof(buffer) - 1, nullptr, String::REPLACE_INVALID_UTF8);
  buffer[sizeof(buffer) - 1] = '\0';
  // Now manipulate the buffer.
}
```